src/triage/component/postmodeling/contrast/model_evaluator.py

```diff
@@ -0,0 +1,1155 @@
+       Model labeled errors (residuals) by the error_labeler (FPR, FNR, and
+       general residual) using a RandomForestClassifier. This function will
+       yield a plot tree for each of the label numpy arrays return by the
+       error_labeler (Y).
+       Arguments:
+           - depth: max number of tree partitions. This is passed directly to
+             the classifier.
+           - view_plot: the plot is saved to disk by default, but the
+             graphviz.Source also allow to load the object and see it in the
+             default OS image renderer
+           - **kwargs: more arguments passed to the labeler: param indicating
+             the threshold value, param_type indicating the type of threshold,
+             and the path to the ProjectStorage.
+       '''
+
+       # Get matrices from the labeler
+       zip_data = self._error_labeler(param_type = kwargs['param_type'],
+                                  param = kwargs['param'],
+                                  path=kwargs['path'])
+
+       # Model tree and output tree plot
+       for error_label, matrix in zip_data:
+
+           dot_path = 'error_analysis_' + \
+                      str(error_label[1]) + '_' + \
+                      str(self.model_id) + '_' + \
+                      str(kwargs['param_type']) + '@'+ \
+                      str(kwargs['param']) +  '.gv'
+
+           clf = tree.DecisionTreeClassifier(max_depth=depth)
+           clf_fit = clf.fit(matrix, error_label[0])
+           tree_viz = tree.export_graphviz(clf_fit,
+                                           out_file=None,
+                                           feature_names=matrix.columns.values,
+                                           filled=True,
+                                           rounded=True,
+                                           special_characters=True)
+           graph = graphviz.Source(tree_viz)
+           graph.render(filename=dot_path,
+                        directory='error_analysis',
+                        view=view_plots)
+
+           print(dot_path)
+
+    def error_analysis(self, threshold_iterator, **kwargs):
+        '''
+        Error analysis function for ThresholdIterator objects. This function
+        have the same functionality as the _error.modeler method, but its
+        implemented for iterators, which can be a case use of this analysis.
+        If no iterator object is passed, the function will take the needed
+        arguments to run the _error_modeler.
+        Arguments:
+            - threshold_iterator: A ThresholdIterator class (see paramters.py)
+            -**kwags: other arguments passed to _error_modeler
+        '''
+
+        error_modeler = partial(self._error_modeler,
+                               depth = kwargs['depth'],
+                               path = kwargs['path'],
+                               view_plots = kwargs['view_plots'])
+
+        if isinstance(threshold_iterator, collections.Iterable):
```

Comment:
not sure this would be preferred, but you can also simply test it:

    if hasattr(threshold_iterator, '__iter__'):
        ...

or __just do it__, _e.g._:

    try:
        threshold_iterator_ = iter(threshold_iterator)
    except TypeError:
        error_modeler(...)
    else:
        for (threshold_type, threshold) in threshold_iterator_:
            ...

Reply:
(on second thought, I'm not sure that your test against `collections.Iterator` will work _at all_ -- for the same reason I was saying it's confusing to call such a thing an iterator -- I'm not sure Python does include `collections.Iterator` in the inheritance hierarchy of `dict` or of classes like yours. the first test I was suggesting, on the other hand, is "duck" type testing whether it's an **iterable**.)